extensions/data-transfer/portability-data-transfer-google/src/main/java/org/dataportabilityproject/datatransfer/google/mail/GoogleMailImporter.java

```diff
@@ -0,0 +1,236 @@
+          newMappingsCreated = true;
+        }
+      }
+    }
+
+    // Persist temp data in case we added mappings along the way
+    if (newMappingsCreated) {
+      jobStore.update(id, tempMailData);
+    }
+
+    // PROCESS MESSAGES
+    for (MailMessageModel mailMessageModel : data.getMessages()) {
+
+      // Gather the label ids that will be associated with this message
+      ImmutableList.Builder<String> importedLabelIds = ImmutableList.builder();
+      for (String exportedLabelIdOrName : mailMessageModel.getContainerIds()) {
+        // By this time all the label ids have been added to tempdata
+        String importedLabelId = tempMailData.getImportedId(exportedLabelIdOrName);
+        if (importedLabelId != null) {
+          importedLabelIds.add(exportedLabelIdOrName);
+        } else {
+          logger.warn(
+              "labels should have been added prior to importing messages"); // TODO remove after
+                                                                            // testing
+        }
+      }
+      // Create the message to import
+      Message newMessage =
+          new Message()
+              .setRaw(mailMessageModel.getRawString())
+              .setLabelIds(importedLabelIds.build());
+      try {
+        gmail.users().messages().insert(USER, newMessage).execute();
+      } catch (IOException e) {
+        return new ImportResult(ResultType.ERROR, "Error importing message: " + e.getMessage());
+      }
+    }
+
+    return new ImportResult(ResultType.OK);
+  }
+
+  /** Supplies a mapping of Label Name -> Label Id (in the import account). */
+  private Supplier<Map<String, String>> allDestinationLabelsSupplier() {
+    return new Supplier<Map<String, String>>() {
+      @Override
+      public Map<String, String> get() {
+        ListLabelsResponse response = null;
+        try {
+          response = gmail.users().labels().list(USER).execute();
+        } catch (IOException e) {
+          throw new RuntimeException("Unable to list labels for user", e);
+        }
+        ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+
+        for (Label label : response.getLabels()) {
+          // TODO: remove system labels
+          builder.put(label.getName(), label.getId());
+        }
+        return builder.build();
+      }
+    };
+  }
+
+  /** Creates the given {@code labelName} in the import service provider and returns the id. */
+  private String createImportedLabelId(String labelName) throws IOException {
+    Label newLabel =
+        new Label()
+            .setName(labelName)
+            .setLabelListVisibility("labelShow")
+            .setMessageListVisibility("show");
+    return gmail.users().labels().create(USER, newLabel).execute().getId();
+  }
+
+  private Gmail getOrCreateGmail(TokensAndUrlAuthData authData) {
+    return gmail == null ? makeGmailService(authData) : gmail;
+  }
+
+  private synchronized Gmail makeGmailService(TokensAndUrlAuthData authData) {
```

Comment:
Same here re Guice. If this is the same code, maybe it could go in a common lib for now?

Reply:
All of the other importers and exporters have this as well, maybe we can move them all into a common library?